File: week2/3_Valid_Parentheses.ts
// Valid Parentheses
// Stack
// Time Complexity: O(n)
// Space Complexity: O(n)

const isValid = (s: string): boolean => {
  let m: string[] = [];
  let brackets = ['(', '{', '['];
  let closed = [')', '}', ']'];

  for (let i = 0; i < s.length; i++) {
    if (closed.includes(s[i])) {
      if (m.length === 0) return false;

      if (m[m.length - 1] === brackets[closed.indexOf(s[i])]) {
        m.pop();
      } else {
        return false;
      }
    } else {
      m.push(s[i]);
    }
  }
  if (m.length === 0) return true;
  return false;
};

// Test cases
console.log(isValid('()'));
console.log(isValid('()[]{}'));
console.log(isValid('(]'));
console.log(isValid('([)]'));

File: week2/3_Valid_Parentheses.py
# Valid Parentheses
# Stack
# Time Complexity: O(n)
# Space Complexity: O(n)

def isValid(s: str) -> bool:
    m = []
    brackets = ['(', '{', '[']
    closed = [')', '}', ']']

    for char in s:
        if char in closed:
            if len(m) == 0:
                return False

            if m[-1] == brackets[closed.index(char)]:
                m.pop()
            else:
                return False
        else:
            m.append(char)

    if len(m) == 0:
        return True
    return False

# Test cases
if __name__ == "__main__":
    print(isValid('()'))
    print(isValid('()[]{}'))
    print(isValid('(]'))
    print(isValid('([)]'))
    print(isValid('{[]}'))
    print(isValid(''))
